File: scripts/seed.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.10"
# dependencies = [
#     "httpx",
# ]
# ///
"""Seed CRM API with challenge dataset: properties, contacts, emails, cases, tasks."""

import json
import os
import sys
from pathlib import Path

import httpx

DATA_PATH = Path(__file__).resolve().parent.parent / "challenge-definition" / "proptech-test-data.json"
CRM_API_URL = os.environ.get("CRM_API_URL", "http://localhost:8002")

client = httpx.Client(base_url=CRM_API_URL, timeout=30)


def api_post(entity: str, data: dict) -> dict:
    resp = client.post(f"/api/{entity}", json=data)
    if resp.status_code != 201:
        print(f"  ERROR creating {entity}: {resp.status_code} {resp.text}", file=sys.stderr)
        resp.raise_for_status()
    return resp.json()


def load_data():
    return json.loads(DATA_PATH.read_text())


def seed_properties(properties):
    """Create a Property per challenge property. Returns {challenge_id: crm_id}."""
    prop_map = {}
    for prop in properties:
        # Derive manager_email from manager name: "John Smith" → "john.smith@manageco.ie"
        manager_name = prop["manager"]
        manager_email = manager_name.lower().replace(" ", ".") + "@manageco.ie" if manager_name else None
        payload = {
            "name": prop["name"],
            "type": prop["type"],
            "units": prop["units"],
            "manager": manager_name,
            "challenge_id": prop["id"],
        }
        if manager_email:
            payload["manager_email"] = manager_email
        created = api_post("properties", payload)
        prop_map[prop["id"]] = created["id"]
        print(f"  Property: {prop['name']} → {created['id']}")
    return prop_map


def extract_senders(emails):
    """Return list of unique senders (first occurrence wins)."""
    seen = {}
    for email in emails:
        addr = email["from"]["email"].lower()
        if addr not in seen:
            seen[addr] = email["from"]
    return list(seen.values())


def split_name(full_name):
    parts = full_name.strip().split()
    if len(parts) == 1:
        return "", parts[0]
    return " ".join(parts[:-1]), parts[-1]


def seed_contacts(senders, prop_map):
    """Create a Contact per unique sender. Returns {email_addr: crm_id}."""
    contact_map = {}
    for sender in senders:
        first, last = split_name(sender["name"])
        payload = {
            "first_name": first,
            "last_name": last,
            "email": sender["email"],
            "type": sender["type"],
        }
        if "role" in sender:
            payload["role"] = sender["role"]
        if "company" in sender:
            payload["company"] = sender["company"]
        if "unit" in sender:
            payload["unit"] = sender["unit"]

        prop_id = sender.get("property_id")
        if prop_id and prop_id in prop_map:
            payload["property_id"] = prop_map[prop_id]

        created = api_post("contacts", payload)
        contact_map[sender["email"].lower()] = created["id"]
        print(f"  Contact: {sender['name']} <{sender['email']}> → {created['id']}")
    return contact_map


def parse_addresses(field):
    if not field:
        return []
    return [a.strip() for a in field.split(",") if a.strip()]


def seed_emails(emails):
    """Create Emails with threading. Returns {challenge_id: crm_id}."""
    sorted_emails = sorted(emails, key=lambda e: (e["thread_id"], e["thread_position"]))
    email_map = {}

    for email in sorted_emails:
        to_addrs = parse_addresses(email["to"])
        cc_addrs = parse_addresses(email.get("cc"))

        payload = {
            "subject": email["subject"],
            "body": email["body"],
            "body_plain": email["body"],
            "from_address": email["from"]["email"],
            "to_addresses": to_addrs,
            "date_sent": email["timestamp"],
            "status": "archived",
            "is_read": email["read"],
            "thread_id": email["thread_id"],
            "thread_position": email["thread_position"],
            "challenge_id": email["id"],
            "message_id": f"<{email['id']}@proptech-challenge>",
        }

        if cc_addrs:
            payload["cc_addresses"] = cc_addrs

        # Link reply to parent via in_reply_to
        if email["thread_position"] > 1:
            parent = next(
                (e for e in sorted_emails
                 if e["thread_id"] == email["thread_id"]
                 and e["thread_position"] == email["thread_position"] - 1),
                None,
            )
            if parent:
                payload["in_reply_to"] = f"<{parent['id']}@proptech-challenge>"

        created = api_post("emails", payload)
        email_map[email["id"]] = created["id"]
        print(f"  Email: {email['id']} \"{email['subject'][:50]}\" → {created['id']}")

    print(f"Seeded {len(sorted_emails)} emails")
    return email_map


def main():
    data = load_data()
    print("Seeding CRM...\n")

    print("Creating Properties...")
    prop_map = seed_properties(data["metadata"]["properties"])
    print(f"\nCreated {len(prop_map)} properties\n")

    print("Creating Contacts...")
    senders = extract_senders(data["emails"])
    contact_map = seed_contacts(senders, prop_map)
    print(f"\nCreated {len(contact_map)} contacts\n")

    print("Creating Emails...")
    seed_emails(data["emails"])

    print("\nSeed complete.")


if __name__ == "__main__":
    main()

File: scripts/seed.ts
/**
 * Seed CRM API with challenge dataset: properties, contacts, emails.
 * Run: npx tsx scripts/seed.ts
 */

import { readFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const DATA_PATH = resolve(__dirname, "../challenge-definition/proptech-test-data.json");
const CRM_API_URL = process.env.CRM_API_URL || "http://localhost:8002";

async function apiPost(entity: string, data: Record<string, unknown>): Promise<Record<string, unknown>> {
  const res = await fetch(`${CRM_API_URL}/api/${entity}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (res.status !== 201) {
    const text = await res.text();
    console.error(`  ERROR creating ${entity}: ${res.status} ${text}`);
    throw new Error(`Failed to create ${entity}`);
  }
  return res.json() as Promise<Record<string, unknown>>;
}

interface ChallengeProperty {
  id: string;
  name: string;
  type: string;
  units: number;
  manager: string;
}

interface ChallengeSender {
  name: string;
  email: string;
  type: string;
  role?: string;
  company?: string;
  unit?: string;
  property_id?: string;
}

interface ChallengeEmail {
  id: string;
  subject: string;
  body: string;
  from: ChallengeSender;
  to: string;
  cc?: string;
  timestamp: string;
  read: boolean;
  thread_id: string;
  thread_position: number;
}

function loadData(): { metadata: { properties: ChallengeProperty[] }; emails: ChallengeEmail[] } {
  return JSON.parse(readFileSync(DATA_PATH, "utf-8"));
}

async function seedProperties(properties: ChallengeProperty[]): Promise<Map<string, number>> {
  const propMap = new Map<string, number>();
  for (const prop of properties) {
    const managerEmail = prop.manager
      ? prop.manager.toLowerCase().replace(/ /g, ".") + "@manageco.ie"
      : undefined;
    const payload: Record<string, unknown> = {
      name: prop.name,
      type: prop.type,
      units: prop.units,
      manager: prop.manager,
      challenge_id: prop.id,
    };
    if (managerEmail) payload.manager_email = managerEmail;
    const created = await apiPost("properties", payload);
    propMap.set(prop.id, created.id as number);
    console.log(`  Property: ${prop.name} → ${created.id}`);
  }
  return propMap;
}

function extractSenders(emails: ChallengeEmail[]): ChallengeSender[] {
  const seen = new Map<string, ChallengeSender>();
  for (const email of emails) {
    const addr = email.from.email.toLowerCase();
    if (!seen.has(addr)) seen.set(addr, email.from);
  }
  return Array.from(seen.values());
}

function splitName(fullName: string): [string, string] {
  const parts = fullName.trim().split(/\s+/);
  if (parts.length === 1) return ["", parts[0]];
  return [parts.slice(0, -1).join(" "), parts[parts.length - 1]];
}

function parseAddresses(field: string | undefined): string[] {
  if (!field) return [];
  return field.split(",").map((a) => a.trim()).filter(Boolean);
}

async function seedContacts(
  senders: ChallengeSender[],
  propMap: Map<string, number>,
): Promise<Map<string, number>> {
  const contactMap = new Map<string, number>();
  for (const sender of senders) {
    const [first, last] = splitName(sender.name);
    const payload: Record<string, unknown> = {
      first_name: first,
      last_name: last,
      email: sender.email,
      type: sender.type,
    };
    if (sender.role) payload.role = sender.role;
    if (sender.company) payload.company = sender.company;
    if (sender.unit) payload.unit = sender.unit;
    if (sender.property_id && propMap.has(sender.property_id)) {
      payload.property_id = propMap.get(sender.property_id);
    }
    const created = await apiPost("contacts", payload);
    contactMap.set(sender.email.toLowerCase(), created.id as number);
    console.log(`  Contact: ${sender.name} <${sender.email}> → ${created.id}`);
  }
  return contactMap;
}

async function seedEmails(challengeEmails: ChallengeEmail[]): Promise<Map<string, number>> {
  const sorted = [...challengeEmails].sort(
    (a, b) => a.thread_id.localeCompare(b.thread_id) || a.thread_position - b.thread_position,
  );
  const emailMap = new Map<string, number>();

  for (const email of sorted) {
    const toAddrs = parseAddresses(email.to);
    const ccAddrs = parseAddresses(email.cc);

    const payload: Record<string, unknown> = {
      subject: email.subject,
      body: email.body,
      body_plain: email.body,
      from_address: email.from.email,
      to_addresses: toAddrs,
      date_sent: email.timestamp,
      status: "archived",
      is_read: email.read,
      thread_id: email.thread_id,
      thread_position: email.thread_position,
      challenge_id: email.id,
      message_id: `<${email.id}@proptech-challenge>`,
    };

    if (ccAddrs.length) payload.cc_addresses = ccAddrs;

    // Link reply to parent via in_reply_to
    if (email.thread_position > 1) {
      const parent = sorted.find(
        (e) => e.thread_id === email.thread_id && e.thread_position === email.thread_position - 1,
      );
      if (parent) payload.in_reply_to = `<${parent.id}@proptech-challenge>`;
    }

    const created = await apiPost("emails", payload);
    emailMap.set(email.id, created.id as number);
    console.log(`  Email: ${email.id} "${email.subject.slice(0, 50)}" → ${created.id}`);
  }

  console.log(`Seeded ${sorted.length} emails`);
  return emailMap;
}

export async function main(): Promise<void> {
  const data = loadData();
  console.log("Seeding CRM...\n");

  console.log("Creating Properties...");
  const propMap = await seedProperties(data.metadata.properties);
  console.log(`\nCreated ${propMap.size} properties\n`);

  console.log("Creating Contacts...");
  const senders = extractSenders(data.emails);
  const contactMap = await seedContacts(senders, propMap);
  console.log(`\nCreated ${contactMap.size} contacts\n`);

  console.log("Creating Emails...");
  await seedEmails(data.emails);

  console.log("\nSeed complete.");
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
